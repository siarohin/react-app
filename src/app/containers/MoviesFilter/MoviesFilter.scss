.app-movies-filter {
  display: grid;
  align-items: center;
  color: var(--app-theme-grey-light-25);
  font-size: 1.8rem;
  grid-template-columns: 1fr minmax(0, max-content);
  grid-template-rows: minmax(6rem, 1fr) var(--app-spacing-xxxs) minmax(6rem, 1fr);
  position: sticky;
  top: 0;
  z-index: 2;
  background: var(--app-theme-grey-light-900);
}

.app-movies-filter__tab-panel {
  grid-row: 1 / span 1;
  grid-column: 1 / span 1;
  height: 100%;
}

.app-movies-filter__select-panel {
  grid-row: 1 / span 1;
  grid-column: 2 / span 1;
  display: inline-grid;
  grid-template-columns: auto 18rem;
  column-gap: var(--app-spacing-xs);
  align-items: center;
}

.app-movies-filter__separator {
  grid-row: 2 / span 1;
  grid-column: 1 / span 2;
  background: var(--app-theme-grey-light-400);
  height: 100%;
}

.app-movies-filter__count {
  grid-row: 3 / span 1;
  grid-column: 1 / span 2;
}

.app-movies-filter__pagination {
  grid-row: 3 / span 1;
  grid-column: 1 / span 2;
  justify-self: center;
}
