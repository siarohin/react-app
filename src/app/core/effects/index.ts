import { combineEpics } from "redux-observable";

import { movieEffects } from "./movies.effect";
import { userPreferencesEffects } from "./user-preferences.effect";
import { toastEffects } from "./toast.effect";

export * from "./common.util";

export const effects = combineEpics(movieEffects, userPreferencesEffects, toastEffects);
