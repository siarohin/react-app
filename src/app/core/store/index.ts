export * from "./reducers";
export { SharedModels } from "./shared";
export { MoviesActions, MoviesModels } from "./movies";
export { UserPreferencesActions, UserPreferencesModels } from "./user-preferences";
export { ToastActions } from "./toast";
