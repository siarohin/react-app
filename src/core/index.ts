export * from "./constants";
export * from "./services";
export * from "./effects";
export * from "./store";
export * from "./store-configuration";
