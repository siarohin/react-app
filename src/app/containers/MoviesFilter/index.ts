import MoviesFilter from "./MoviesFilter";

export { MoviesFilter };
