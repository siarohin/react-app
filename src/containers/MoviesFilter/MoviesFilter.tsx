import React, { ReactElement } from "react";

import "./MoviesFilter.scss";
import { FilterButtons, FilterCount, FilterSelect } from "../../components";
import { IGenresFilter } from "../../models";

export const MoviesFilter = (props: IGenresFilter): ReactElement => {
  return (
    <div className="app-movies-filter">
      <div className="app-movies-filter__tab-panel">
        <FilterButtons {...props} />
      </div>
      <div className="app-movies-filter__select-panel">
        <FilterSelect changeSorting={props.changeSorting} sortingOptions={props.sortingOptions} />
      </div>
      <div className="app-movies-filter__count">{props.isLoading ? null : <FilterCount count={props.count} />}</div>
      <div className="app-movies-filter__separator"></div>
    </div>
  );
};
