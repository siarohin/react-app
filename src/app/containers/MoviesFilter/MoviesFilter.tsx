import React, { ComponentType, ReactElement } from "react";
import { connect } from "react-redux";

import "./MoviesFilter.scss";
import { FilterButtons, FilterCount, FilterSelect } from "../../components";
import { SharedModels, UserPreferencesModels, UserPreferencesActions, MoviesModels, State } from "../../core";
import { PaginationLayout } from "../PaginationLayout";

const MoviesFilter = (
  props: Partial<MoviesModels.IMoviesState> & UserPreferencesModels.IUserPreferencesState & SharedModels.IDispatchAction
): ReactElement => {
  const { totalAmount, genres, sortingOptions, dispatch } = props;

  const handleSelectedGenre = (selected: string) => {
    if (genres.selected !== selected) {
      dispatch(UserPreferencesActions.updateSelectedGenre({ selected }));
    }
  };

  const handleChangeSorting = (selected: string) => {
    if (sortingOptions.selected !== selected) {
      dispatch(UserPreferencesActions.updateSelectedSortingOption({ selected }));
    }
  };

  return (
    <div className="app-movies-filter">
      <div className="app-movies-filter__tab-panel">
        <FilterButtons selected={handleSelectedGenre} genres={genres} />
      </div>
      <div className="app-movies-filter__select-panel">
        <FilterSelect changeSorting={handleChangeSorting} sortingOptions={sortingOptions} />
      </div>
      <div className="app-movies-filter__count">{totalAmount ? <FilterCount count={totalAmount!} /> : null}</div>
      <div className="app-movies-filter__pagination">{<PaginationLayout />}</div>
      <div className="app-movies-filter__separator"></div>
    </div>
  );
};

const mapStateToProps = (
  state: State
): Partial<MoviesModels.IMoviesState> & Partial<UserPreferencesModels.IUserPreferencesState> => {
  return {
    totalAmount: state.movieList.totalAmount,
    genres: state.userPreferences.genres,
    sortingOptions: state.userPreferences.sortingOptions
  };
};

export default connect(mapStateToProps)(MoviesFilter as ComponentType<any>);
