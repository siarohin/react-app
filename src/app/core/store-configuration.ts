import { applyMiddleware, createStore, compose, Store } from "redux";
import { combineEpics, createEpicMiddleware, Epic } from "redux-observable";

import { reducers, State } from "./store";
import { isServer } from "./constants";

function getComposeEnhancers(): any {
  if (!isServer) {
    return (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
  }
  return compose;
}

export interface AppStore {
  store: Store;
  runEffects: (effects$?: Array<Epic>) => void;
}

/**
 * Store configuration
 */
export function configureStore(initialState: State = {} as State): AppStore {
  const epicMiddleware = createEpicMiddleware();
  const composeEnhancers = getComposeEnhancers();
  const middlewares = [epicMiddleware];

  const store: Store = createStore(reducers, initialState, composeEnhancers(applyMiddleware(...middlewares)));

  return {
    store,
    runEffects: (effects$?: Array<Epic>) =>
      effects$
        ? epicMiddleware.run(combineEpics(...effects$))
        : // eslint-disable-next-line no-console
          console.warn("Effects are not running, no epics array was provided")
  };
}
