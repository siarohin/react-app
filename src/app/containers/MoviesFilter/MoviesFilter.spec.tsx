import React from "react";
import { Provider } from "react-redux";
import renderer, { ReactTestRenderer } from "react-test-renderer";
import configureStore from "redux-mock-store";

import { State } from "../../core";
import MoviesFilter from "./MoviesFilter";

const mockStore = configureStore([]);

describe("Containers.MoviesFilter.MoviesFilter: ", () => {
  let store: any;
  let component: ReactTestRenderer;

  beforeEach(() => {
    store = mockStore({
      movieList: {
        totalAmount: 3000
      },
      userPreferences: {
        genres: {
          all: ["Romantic"],
          selected: ""
        },
        sortingOptions: {
          options: ["Rating", "Release date"],
          selected: "Release date"
        }
      }
    } as State);

    component = renderer.create(
      <Provider store={store}>
        <MoviesFilter />
      </Provider>
    );
  });

  it("should render component", () => {
    expect(component.toJSON()).toMatchSnapshot();
  });
});
