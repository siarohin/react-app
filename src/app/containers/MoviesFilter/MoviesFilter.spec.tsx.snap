// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Containers.MoviesFilter.MoviesFilter:  should render component 1`] = `
<div
  className="app-movies-filter"
>
  <div
    className="app-movies-filter__tab-panel"
  >
    <div
      aria-label="text alignment"
      className="MuiToggleButtonGroup-root app-filter-buttons"
      role="group"
    >
      <button
        aria-pressed={true}
        className="MuiButtonBase-root MuiToggleButton-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal app-filter-buttons__toggle-button Mui-selected"
        disabled={false}
        onBlur={[Function]}
        onChange={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
        value="All"
      >
        <span
          className="MuiToggleButton-label"
        >
          All
        </span>
      </button>
      <button
        aria-pressed={false}
        className="MuiButtonBase-root MuiToggleButton-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal app-filter-buttons__toggle-button"
        disabled={false}
        onBlur={[Function]}
        onChange={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
        value="Romantic"
      >
        <span
          className="MuiToggleButton-label"
        >
          Romantic
        </span>
      </button>
    </div>
  </div>
  <div
    className="app-movies-filter__select-panel"
  >
    <div
      className="app-filter-select__prefix"
    >
      Sort by
    </div>
    <div
      className="MuiFormControl-root app-filter-select__control"
    >
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline app-filter-select__select MuiInputBase-formControl MuiInput-formControl"
        onClick={[Function]}
      >
        <div
          aria-haspopup="listbox"
          className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="button"
          tabIndex={0}
        >
          Rating
        </div>
        <input
          aria-hidden={true}
          className="MuiSelect-nativeInput"
          onAnimationStart={[Function]}
          onChange={[Function]}
          required={false}
          tabIndex={-1}
          value="Rating"
        />
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSelect-icon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
  </div>
  <div
    className="app-movies-filter__count"
  >
    <span
      className="app-filter-count"
    >
      3000
    </span>
     movies found
  </div>
  <div
    className="app-movies-filter__pagination"
  />
  <div
    className="app-movies-filter__separator"
  />
</div>
`;
